### ISSUE TITLE
Focus advice for content-heavy dialogs

### ISSUE THREAD

**Issue by patrickhlauke** [w3c/aria-practices#442]
While the advice given for dialogs https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal makes sense for "actionable" dialogs which have a relatively short description / prose and whose main purpose is to provide the user with form controls (e.g. some settings, an OK/CANCEL set, etc), dialogs are often used  as informational/content-rich overlays that contain structured content - for instance, a dialog which then holds a lot of text with headings, bulleted lists, etc, and that are used to present information to the user.

In these cases, although the described pattern can of course be used (tie up all the content to the dialog element using `aria-describedby`, set focus on the close or ok control, etc), doing so may not be ideal for SR users as the structured content will be announced without the immediate option to properly use the reading keys to read/explore it.

For these types of dialogs, I wonder if focus should be allowed to go to the dialog container or the first heading/start of the content (as would happen with in-page links jumping to a part of the current document), rather than to the first focusable element?

**Comment by patrickhlauke** [comment]
To clarify (as @heydon rightly reminded me) when I say "setting focus to the heading/start of content" i mean using `tabindex="-1"`, not `tabindex="0"`

**Comment by StommePoes** [comment]
That's what I've been pushing my colleagues to do. I want to have the main dialog patterns more for alertdialogs-- they're always short, sweet, and follow the desktop/application dialog pattern well. Those also don't need role=document inside for NVDA users who don't realise some web designer basically threw a whole document inside a modal (whole other topic).

**Comment by MichielBijl** [comment]
To summarise: if dialog is content heavy set focus to the content area with tabindex="-1"?

**Comment by StommePoes** [comment]
Michiel: we mean the actual role=dialog element itself, rather than someone inside near the top (often we have a close button as the first child and it's already focusable so with current code we've been focussing on this, since users don't get skipped over buttloads of text if they go "forward" from there).

Maybe spec-wise it doesn't matter who we focus on but if you're building reusable components it's nice to be able to say "focus on element x each time" since often our modals have no clue what ungodly steaming pile of content might be stuffed inside.

**Comment by patrickhlauke** [comment]
so from where i'm standing (well, sitting) i think the safest, most generic advice would be: set focus to the container with `role="dialog" tabindex="-1"`; however, IF your dialog only contains a short explanatory text (which you're associating correctly already to the dialog via `aria-labelledby`/`aria-describedby`/etc), then it's fine to focus on the first focusable/actionable item inside the dialog. if that makes sense / isn't too vague. and then give examples of a content-heavy dialog (used as a popup to show structured content), where focus moves to the `role="dialog"` container, versus the other type of dialog (short title, short paragraph of description, some form controls/buttons).

**Comment by mcking65** [comment]
Please do not encourage people to place focus on the dialog container.

Per the guidance the task force has provided, if the dialog content is long, and If there is a close button at the top, focusing on close is a good choice. Otherwise, if not present, focus on a contained static element near the top.

Keep in mind that a dialog is a window, not a region. this has important implications. When in reading mode, boundaries are not part of the reading view like they are with inline container elements like regions.

The typical screen reader algorithm for a focus change that moves focus into a new dialog is:
1. Read dialog label and role
2. Read dialog description
3. Read focused element

With most screen readers, the reading cursor tracks the focus. So, after reading the focused element, the screen reader reading cursor is placed on that element.

Thus, if you focus on a close button at the top of a long dialog, then the user will hear:
1. Title Dialog
2. Descriptive text
3. Close button

And the reading cursor will be on the close button, which the user could then either activate or use as a logical starting point for re-reading what was just heard. This is a clean, understandable experience that does not have any confusing or annoying repetition.

If you instead place focus on the dialog window, the user will likely hear everything twice and the location of the reading cursor is not predictable. In addition, screen reader responses to current focused item will be long and confusing, reading off multiple focusable elements. If the user then tries to understand the focus by moving the reading cursor to the focused element, if it works at all, it will probably go to the title or close element, which will not actually be the same as the focus.

We can avoid such confusion without asking screen reader developers to change their conventions by following the guidance the task force has drafted.

Given this issue, does the Authoring Practices treatment of this topic lack sufficient prominence or clarity?

**Comment by patrickhlauke** [comment]
@mcking65 

> Per the guidance the task force has provided, if the dialog content is long, and If there is a close button at the top, focusing on close is a good choice. Otherwise, if not present, focus on a contained static element near the top

Unless I'm misreading the guidance, I don't see anything that can be interpreted as the "otherwise" clause you mention.

> if you focus on a close button at the top of a long dialog, then the user will hear:
>
> 1. Title Dialog
> 2. Descriptive text
> 3. Close button
>


But the problem here is that if the "descriptive text" is complex (contains headings, lists, etc), it is read in one big chunk, not allowing for proper reading keys navigation (as you're focused on the close button at this stage). Also, users may not be aware that they can proceed past that close button they're currently focused on to get to the actual content that was read out in one big go (and, in NVDA, at least older versions, they were prevented from using reading keys when in a `role="dialog"`).

> And the reading cursor will be on the close button, which the user could then either activate or use as a logical starting point for re-reading what was just heard. This is a clean, understandable experience that does not have any confusing or annoying repetition.

I'm not entirely sure if being on the close button sounds like a logical *starting* point. through habit, SR users would probably know/guess that the close button is at the top and precedes the actual content, but coming to this completely fresh it does feel odd to say that being on "close", after having heard all content (assuming it's all tied to the dialog correctly), they can go forward to read that content again with proper reading keys.

Particularly if the content/"descriptive text" of the dialog is in fact long and complex, users will invariably get repetition if they're attempting to read/navigate that complex content in a more structured way (rather than getting it as one continuous stream where they can't go back, skip ahead, etc).

But I understand the concern of not setting focus to the dialog itself. So how about expanding the bullet in the note

> Unless a condition where doing otherwise is advisable, focus is initially set on the first focusable element.

to give an example of a dialog whose purpose is to convey complex structured content, and to suggest that in those situations, it is possible to set focus on the first static element of that content? depending on structure, this would skip the close button, but users would be able to reach it by tabbing or reverse tabbing (assuming that focus is contained in the dialog, as it should)?

**Comment by MichielBijl** [comment]
@patrickhlauke 

> Unless I'm misreading the guidance, I don't see anything that can be interpreted as the "otherwise" clause you mention.

It’s in the paragraph I posted earlier. It’s the second bullet under list item 1 in the note under “Keyboard Interaction”.

**Comment by patrickhlauke** [comment]
That paragraph talks about content that leads to scrolling (which is problematic, as scrolling depends on the size of the user's window/viewport, so what scrolls for one author/auditor may not scroll for another etc). If that can be generalised more (not about scrolling, or at least not just about scrolling, but also about structural complexity), then perhaps yes this can be addressed this way.

**Comment by StommePoes** [comment]
>If you instead place focus on the dialog window, the user will likely hear everything twice and the location of the reading cursor is not predictable.

Is this true even if we correctly aria-labelledby the dialog with the heading and aria-describedby with a (small chunk of or summary!) the text inside? Instead the whole dialog will be read out? Because we already have that problem now if/when the screen reader treats the dialog like a desktop application dialog. 

I say "small chunk of, or summary", because otherwise in our cases it will very much be no different from wrapping all the content of a web page to aria-describe the body element. Designers love modals and the more they see Large Companies (cough*google*cough) putting everything into modals, the more they see it as okay. We have iframes inside our modals (because our forms are using TinyMCE editables). We have things launching alertdialogs and confirmation dialogs (are you sure? and Warning! This will delete user data!) and can launch desktop windows such as image/file upload.

If we should focus on the first static element, then our best hope is that the consumer of our modal has always given it a heading and have our code look for the heading to focus on (we don't always have a close button, but when we do we focus on that).

**Comment by devarshipant** [comment]
We test a lot of modal dialogs and I have noticed the spoken output with SR is more meaningful when focus is set on the dialog, aria-labelledby (on parent div) references the dialog title (heading), and aria-label (on child) complements the dialog. 

Important to note -- aria attributes (labelledby and aria-label) should not be on the same div as only one gets passed on to SR. Note that aria-describedby causes undesirable verbosity.

An example with correct placement of aria attributes -- https://assets.cms.gov/resources/framework/2.0/Pages/modal.html

**Comment by patrickhlauke** [comment]
it would be good to revisit/address this at some point, as i'm seeing an increased use of dialogs that are content-heavy, or contain complex mixed content/forms/controls, that still adopt the nominal pattern that is not really well suited to them

**Comment by mcking65** [comment]
@patrickhlauke

Currently, the first item in the keyboard guidance section is:
>* When a dialog opens, focus moves to an element inside the dialog. See notes below regarding initial focus placement.

The second item in the notes is:

>* Unless a condition where doing otherwise is advisable, focus is initially set on the first focusable element. 

The third item is:

>* If content is large enough that focusing the first interactive element could cause the beginning of content to scroll out of view, it is advisable to add tabindex= " -1 "  to a static element at the top of the dialog, such as the dialog title or first paragraph, and initially focus that element. 

I am assuming you are suggesting we revise that 3rd note? In particular, the description of the advisable condition is insufficient. It only says to do this if the content is large enough that the first focusable element will scroll out of view.

Often there is a small bit of static text, a short sentence or phrase, that precedes an input at the top of a dialog. In such circumstances, focusing the input is very desireable for sighted keyboard users, and screen reader users are still well served if that static text is referenced by aria-describedby. In this scenario, it seems we wouldn't want to suggest focusing a static element, forcing keyboard users to tab to the input.

Do you have a suggestion for how we would rephrase that 3rd note? If necessary, we can replace it with a short nested list of conditions.

**Comment by patrickhlauke** [comment]
Revisiting what I said earlier here - just checking that things hadn't changed since then:

> how about expanding the bullet in the note
>
> > Unless a condition where doing otherwise is advisable, focus is initially set on the first focusable element.
>
> to give an example of a dialog whose purpose is to convey complex structured content, and to suggest that in those situations, it is possible to set focus on the first static element of that content?

> That paragraph talks about content that leads to scrolling (which is problematic, as scrolling depends on the size of the user's window/viewport, so what scrolls for one author/auditor may not scroll for another etc). If that can be generalised more (not about scrolling, or at least not just about scrolling, but also about structural complexity), then perhaps yes this can be addressed this way.

Because the only other "condition" that the note explains is the case of long content that would lead to scrolling. But with complex content/structures/dialogs that are essentially conveying information rather than being just a title, short desc, and a handful of form controls, it's not necessarily any scrolling that's the problem, but the fact that content just isn't conveyed appropriately if it's all vomited out in one big go thanks to `aria-describedby`.

So, don't know, as a starting point, change that second bullet point from: `Unless a condition where doing otherwise is advisable, focus is initially set on the first focusable element.` to `Unless a condition where doing otherwise is advisable, focus is initially set on the first focusable element. Conditions can include:` and then make the current third bullet the first bullet of a sublist (about the scrolling), then add a second sub-bullet along the rough lines of

```
If the dialog contains complex/structured content, and/or its primary aim is to convey content (rather than a short confirmation dialog, or a form),we recommend *not* associating the content with the dialog using aria-describedby, as this will result in AT reading the entirety of the content in one unbroken (and non-structured) stream to the user. Instead, do not use aria-describedby, and place focus to the start of the content. This way, AT users can consume the structured content as they would any other content in the page, using their reading keys...
```

or something to that effect.

**Comment by patrickhlauke** [comment]
Made a first stab at addressing this (as well as the issue of `aria-describedby` being used inappropriately on content/structure-heavy dialogs) here #1707
